Doc/library/ipaddress.rst

```diff
@@ -311,15 +322,17 @@ IPv6).
 Comparison operators
 """"""""""""""""""""
 
-Address objects can be compared with the usual set of comparison operators.  Some
-examples::
+Address objects can be compared with the usual set of comparison operators. Scope id
+is not considered while comparing IPv6 address objects. Some examples::
```

Comment:
Ah, is this the behavior we want?

I think probably not. In particular the whole point of this scoped-address system, if I'm understanding it right, is that the address may be local to a particular scope -- so that both link 1 and link 2 might have an address called `fe80::1234`, and those would be completely unrelated addresses that might belong to entirely different machines.

So we really want `IPv6Address('fe80::1234%1') != IPv6Address('fe80::1234%2')`.

Reply:
While writing code, I investigated this issue and came to the decision, that same IPv6 addresses with different zone id (or scope id) should be considered equal. Unfortunately, can not provide RFC or other 100% proof, but in [open-jdk](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/share/classes/java/net/Inet6Address.java#L269-L276) zone id (scope id) is not considered while comparing IPv6 addresses. Also, [stackowerflow discussion](https://stackoverflow.com/questions/1787530/how-do-you-determine-equality-between-two-ipv6-addresses).

Reply:
> Unfortunately, can not provide RFC or other 100% proof, but in open-jdk zone id (scope id) is not considered while comparing IPv6 addresses.

@opavlyuk Is there anything else besides open-jdk which also considers two of the same IPv6 addresses to be equal even when their scope IDs differ?

Personally, I think it's reasonable for this to be a separate PR and/or issue. At the moment, there is no support at all for scoped IPv6 addresses (as far as I'm aware) within stdlib. In it's current state, this PR provides a significant improvement.

Reply:
Found out, that in [dotnet core](https://github.com/dotnet/corefx/blob/master/src/System.Net.Primitives/src/System/Net/IPAddress.cs#L553-L578) scope (zone) id **is considered** while comparing IPv6 addresses.

Need help with the comparison issue, because can not find any reputable sources.

Reply:
@opavlyuk If there's no response after a decent period in the PR itself, it might be worthwhile to open a new topic over on Python-Dev to see what the others think the behavior should be.

Reply:
Some person's opinion on StackOverflow is not a particularly reliable source for deciding the right way to handle something. :smile: It's often a good source of links, or names of things to go read up on in the documentation/standards/etc... but there are a *lot* of bad and buggy suggestions there.

That leaves us, at the end of the day, responsible for our own reading of the underlying standard and other technical context.

And what I see from that reading is: The entire point of RFC 4007 -- of the whole system of scoped addresses, zones, and writing down as part of an address some identifier for a scope zone -- is that the same 128-bit address can mean different things in different zones. The zone identifier is there to distinguish those different meanings.

The [Wikipedia article](https://en.wikipedia.org/wiki/IPv6_address#Scoped_literal_IPv6_addresses) (following a link from that SO thread! just like I mentioned :wink: ) puts this nicely:

> For addresses with other than global scope [...], and in particular for link-local addresses [...] the same address may be valid in different zones, and be in use by a different host in each of those zones.

(There's nothing new there -- the RFC says all the same things, just with more words.)

Given that the whole point of the system is to give different meanings to the same 128 bits in different zones, the only consistent semantics for `==` on such things is that the same 128-bit address plus two different zone IDs makes two distinct, not-equal values.


> Found out, that in dotnet core scope (zone) id is considered while comparing IPv6 addresses.

Cool. This is helpful for confirming that we won't somehow face interoperability problems where the whole rest of the world turns out to have settled on a shared buggy behavior here.


> I think it's reasonable for this to be a separate PR and/or issue.

I have to disagree with this. Sometimes it's acceptable to introduce broken semantics in a new area, and plan to fix it up later, when implementing the right semantics is a lot of work... but that certainly isn't the case here. We shouldn't introduce the buggy behavior in the first place.

Reply:
> I have to disagree with this. Sometimes it's acceptable to introduce broken semantics in a new area, and plan to fix it up later, when implementing the right semantics is a lot of work... but that certainly isn't the case here. We shouldn't introduce the buggy behavior in the first place.

The distinction between "buggy behavior" vs limited comparison accuracy is not always immediately obvious, that's why I was asking what behavior other platforms utilized. If it is only open-jdk which considers two of the same IPv6 addresses with differing scopes to be equal, it seems to lean a bit more towards being a bug. 

Also, as you mentioned, interoperability can also be a concern. Based on the information gathered since then, it seems that won't be an issue for the most part.

If the comparison can be fixed with a reasonable amount of effort within the same PR and the majority of other platforms consider the scopes during comparisons, I agree that the behavior should be adjusted accordingly.